kafka-instrumentation/src/main/java/dev/responsive/kafka/clients/ResponsiveKafkaClientSupplier.java

```diff
@@ -0,0 +1,276 @@
+ * limitations under the License.
+ */
+
+package dev.responsive.kafka.clients;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import org.apache.kafka.clients.CommonClientConfigs;
+import org.apache.kafka.clients.admin.Admin;
+import org.apache.kafka.clients.consumer.Consumer;
+import org.apache.kafka.clients.producer.Producer;
+import org.apache.kafka.common.Configurable;
+import org.apache.kafka.common.metrics.JmxReporter;
+import org.apache.kafka.common.metrics.KafkaMetricsContext;
+import org.apache.kafka.common.metrics.MetricConfig;
+import org.apache.kafka.common.metrics.Metrics;
+import org.apache.kafka.common.metrics.MetricsContext;
+import org.apache.kafka.common.metrics.MetricsReporter;
+import org.apache.kafka.common.utils.Time;
+import org.apache.kafka.streams.KafkaClientSupplier;
+import org.apache.kafka.streams.StreamsConfig;
+import org.apache.kafka.streams.processor.internals.DefaultKafkaClientSupplier;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * An implementation of KafkaClientSupplier for Responsive applications. The supplier returns
+ * wrapped kafka producers/consumers that are instrumented to work with the responsive platform
+ * by, for example, emitting metrics useful for debug and scaling.
+ *
+ * Synchronization: This class is read/written during client initialization and close (via the
+ * close callbacks on the returned clients). As these calls are not on the performance path we
+ * rely on coarse-grained locks around sections reading/writing shared state.
+ */
+public final class ResponsiveKafkaClientSupplier implements KafkaClientSupplier, Configurable {
+  private static final Logger LOGGER = LoggerFactory.getLogger(ResponsiveKafkaClientSupplier.class);
+  private final KafkaClientSupplier wrapped;
+  private final Factories factories;
+  private boolean eosV2 = false;
+  private boolean configured = false;
+  private Metrics metrics;
+  private EndOffsetsPoller endOffsetsPoller;
+
+  private final Map<String, ReferenceCounted<MetricPublishingCommitListener>> metricListeners
+      = new HashMap<>();
+
+  public ResponsiveKafkaClientSupplier() {
+    this(new Factories() {}, new DefaultKafkaClientSupplier());
+  }
+
+  ResponsiveKafkaClientSupplier(final Factories factories, final KafkaClientSupplier wrapped) {
+    this.factories = factories;
+    this.wrapped = wrapped;
+  }
+
+  @Override
+  public void configure(final Map<String, ?> configs) {
+    LOGGER.info(
+        "Configuring the client supplier. Got configs: {}",
+        configs.entrySet().stream()
+            .map(e -> e.getKey() + ":" + e.getValue())
+            .collect(Collectors.joining("\n"))
+    );
+    eosV2 = (configs.get(StreamsConfig.PROCESSING_GUARANTEE_CONFIG))
```

Comment:
Should we check for eos v1 just in case?

Reply:
I guess there's nothing here that wouldn't work for v1. I'll add that.